src/dd/dd.rs

```diff
@@ -0,0 +1,745 @@
+                        Ok(m) => to_skip -= m,
+                        _ => err("cannot seek on output file".to_string()),
+                    }
+                }
+                if to_skip > 0 {
+                    while to_skip > 0 {
+                        match f.write(&zero_buf[..cmp::min(to_skip, BUF_SIZE)]) {
+                            Ok(0) => err("cannot seek on output file".to_string()),
+                            Ok(m) => to_skip -= m,
+                            _ => err("cannot seek on output file".to_string()),
+                        }
+                    }
+                }
+            }
+        }
+
+        // Would have terminated early if seek != 0, test precondition anyway
+        OutFileType::Stdout(_) => assert_eq!(cf.seek, 0),
+    }
+    (inp, out)
+}
+
+fn copy(
+    mut inp: InpFileWrap,
+    mut out: OutFileWrap,
+    cf: Arc<Mutex<Conf>>,
+    _sdone: chan::Sender<()>,
+) -> result::Result<(), String> {
+    let first = Arc::clone(&cf);
+    let mut cf = cf.lock().unwrap();
+    let mut buf = vec![0u8; cmp::min(cf.inp_block_size, cf.out_block_size)];
+    let mut total = 0;
+    for _ in 0..cf.count {
+        let mut pos = 0;
+        let mut left = cf.inp_block_size;
+        let mut session = 0;
+        'read_block: loop {
+            if pos == cf.inp_block_size {
+                break 'read_block;
+            }
+            let mut is_read_0 = false;
+            match inp.read(&mut buf[pos..pos + left]) {
+                Ok(0) => is_read_0 = true,
+                Ok(num) => {
+                    session += num;
+                    total += num;
+                    if num < cf.inp_block_size {
+                        cf.inp_part += 1;
+                        pos += num;
+                        left = cf.inp_block_size - pos;
+                    } else if num == cf.inp_block_size {
+                        cf.inp_full += 1;
+                    } else {
+                        err(format!("overread: {} over {}", num, cf.inp_block_size));
+                    }
+                }
+                Err(_) => {
+                    // TODO: handle
+                }
+            }
+            if session == cf.inp_block_size || is_read_0 {
+                break 'read_block;
+            }
+        }
+        let num = cmp::min(session, cf.inp_block_size);
+        let mut pos = 0;
+        loop {
+            if pos == num {
+                break;
+            }
+            if let Ok(m) = out.write(&buf[pos..pos + num]) {
+                if m < cf.out_block_size {
+                    cf.out_part += 1;
+                } else {
+                    cf.out_full += 1;
+                }
+                pos += m;
+            } else {
+                err("error occured".to_string()); // TODO: handle
+            }
+        }
+        if total >= cf.inp_block_size * cf.count {
+            break;
+        }
+    }
+    drop(cf);
+    display_stat(Arc::clone(&first));
```

Comment:
This clone isn't necessary.

Reply:
Sadly, I don't know how to do it another way. I clone because its also used signal handler,
so I don't pass it by value.

Reply:
Maybe  you will share an idea?

Reply:
You cloned it earlier so you should be able to just pass ownership of the `Arc` to `display_stat()` (unless this causes a compiler error for you?  If it does let me know so I can look more in-depth).